topdown/reachable.go

```diff
@@ -0,0 +1,54 @@
+package topdown
+
+import (
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+func builtinReachable(graphVal, initialVal ast.Value) (ast.Value, error) {
+	graph, err := builtins.ObjectOperand(graphVal, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	initial, err := builtins.SetOperand(initialVal, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// This is a queue that holds all nodes we still need to visit.  It is
+	// initialised to the initial set of nodes we start out with.
+	queue := []*ast.Term{}
+	initial.Foreach(func(t *ast.Term) {
+		queue = append(queue, t)
+	})
+
+	// This is the set of nodes we have reached.
+	reached := ast.NewSet()
+
+	// Keep going as long as we have nodes in the queue.
+	for len(queue) > 0 {
+		// Get the edges for this node.  If the node was not in the graph,
+		// `edgesTerm` will be `nil` and we can ignore it.
+		node := queue[0]
+		if edgesTerm := graph.Get(node); edgesTerm != nil {
+			// Add all the newly discovered neighbors.
+			if edgeSet, ok := edgesTerm.Value.(ast.Set); ok {
```

Comment:
🤔 I'm wondering what behavior we want if someone supplies an "invalid" object, like in this case what if this isn't a set? Should we maybe just bail out and have the result be undefined? My concern would be that we return a set of results that wasn't what the user had intended and might be hiding the malformed object value. Not sure how big of a deal that is though..

Reply:
I'd keep the current implementation--we can add another that is more strict in the future if needed. I'm thinking of scenarios where the data uses `null` to signal an empty collection. Just my 2c.

Reply:
Works for me 👍